Translation of the 'useDeferredValue' page (#484)
reactjs/fr.react.dev: src/content/reference/react/useDeferredValue.md

}
```

Durant le rendu initial, la <CodeStep step={2}>valeur délayée</CodeStep> sera la même que la <CodeStep step={1}>valeur</CodeStep> que vous avez donné.

Pendant la mise à jour, la <CodeStep step={2}>valeur délayée</CodeStep> va "être en retard" par rapport à la dernière <CodeStep step={1}>valeur</CodeStep>. Plus particulièrement, React va d'abord faire un rendu *sans* mettre à jour la valeur délayée, et puis essayer un autre rendu avec la nouvelle valeur reçue en arrière-plan.

**Essayons un exemple afin de comprendre l'utilité de ce Hook.**

<Note>

Cet exemple part du principe que vous utilisez une de ces méthodes intégrées avec Suspense : 

- Des frameworks de récupération de données utilisant Suspense comme [Relay](https://relay.dev/docs/guided-tour/rendering/loading-states/) ou [Next.js](https://nextjs.org/docs/advanced-features/react-18)

> - Des composants de Lazy-loading (*chargement fainéant, NdT*) avec [`lazy`](/reference/react/lazy)

Reviewer: Ici y'avait glissement sémantique : ce ne sont pas les composants qui font du lazy loading, on parle du fait de charger à la demande les composants (en utilisant `React.lazy` en l'occurrence). Par ailleurs, on a une trad FR établie ("chargement à la demande"), donc pas besoin de garder "lazy loading".

```js SearchResults.js hidden
import { fetchData } from './data.js';

> // Note: ce composant est écrit en utilisant une API expérimentale
> // qui n'est pas encore disponible dans les versions stables de React.

Reviewer: ```suggestion
// Remarque : ce composant est écrit en utilisant une API expérimentale
// qui n'est pas encore disponible dans les versions stables de React.
```

// Si vous souhaitez suivre cet exemple avec une version stable, 
// essayez un framework intégrant Suspense, comme Relay ou Next.js.

export default function SearchResults({ query }) {
  if (query === '') {
    return null;
  }
  const albums = use(fetchData(`/search?q=${query}`));
  if (albums.length === 0) {

>     return <p>Pas de résultats pour <i>"{query}"</i></p>;

Reviewer: ```suggestion
    return <p>Pas de résultat pour <i>« {query} »</i></p>;
```

  );
}

> // Ceci est une astuce permettant de contourner un bug afin de lancer la démo. 
> // TODO: remplacer avec la vraie implémentation quand le bug sera fixé. 

Reviewer: ```suggestion
// Astuce de contournement d'un bug afin d'exécuter la démo. 
// TODO: remplacer avec la véritable implémentation quand le bug sera corrigé.
```

<DeepDive>

> #### Comment délayer une valeur fonctionne-t-il en sous-jacent ? {/*how-does-deferring-a-value-work-under-the-hood*/}

Reviewer: ```suggestion
#### Comment une valeur différée fonctionne-t-elle sous le capot ? {/*how-does-deferring-a-value-work-under-the-hood*/}
```

Reviewer: Tiens, qu'est-ce qui te gênait ici ? "Sous le capot" est la trad établie (et littérale) de "under the hood", et on l'utilise régulièrement dans ces docs (y compris l'édition précédente).

> Vous pouvez imaginer le déroulement en deux étapes : 

Reviewer: ```suggestion
Imaginez un déroulement en deux étapes :
```

> 1. **Premièrement, React re-renders avec la nouvelle `query` (`"ab"`) mais avec l'ancienne `deferredQuery` (toujours `"a")`.** La valeur `deferredQuery`, à laquelle vous passez la liste de résultats, est *délayée:* elle "se met en retard" par rapport à la valeur `query`.

Reviewer: ```suggestion
1. **Pour commencer, React refait un rendu avec la nouvelle `query` (`"ab"`) mais avec l'ancienne `deferredQuery` (toujours `"a")`.** La valeur `deferredQuery`, que vous passez à la liste de résultats, est *différée* : elle est « en retard » par rapport à la valeur `query`.
```

> 2. **En arrière-plan, React essaye de re-render avec `query` et `deferredQuery` mis à jour *ensemble* avec `"ab"`.** Si ce re-render réussi, React vous le montre sur l'écran. Cependant, s'il se suspend (les résultats pour `"ab"` n'ont pas encore été chargés), React va abandonner cet essai de rendu, et essayer de re-render après que les données aient été chargées. L'utilisateur va continuer à voir l'ancienne valeur reportée jusqu'à ce que les données soient chargées. 

Reviewer: ```suggestion
2. **En arrière-plan, React tente alors un autre rendu avec `query` et `deferredQuery` valant *toutes les deux* `"ab"`.** Si ce rendu aboutit, React l'affichera à l'écran. Cependant, s'il suspend (les résultats pour `"ab"` ne sont pas encore chargés), React abandonnera cet essai de rendu, et essaiera à nouveau une fois les données chargées. L'utilisateur continuera à voir l'ancienne valeur différée jusqu'à ce que les données soient prêtes. 
```

> Le rendu de "l'arrière-plan" délayé est susceptible d'être interrompu. Par exemple, si vous tapez dans la zone de saisie à nouveau, React va l'abandonner et recommencer avec la nouvelle valeur. React va toujours utiliser la dernière valeur donnée. 

Reviewer: ```suggestion
Le rendu différé « d'arrière-plan » est susceptible d'être interrompu. Par exemple, si vous tapez à nouveau dans le champ de saisie, React l'abandonnera et recommencera avec la nouvelle valeur. React utilisera toujours la dernière valeur fournie.
```

> Notez qu'il y a toujours une requête réseau par frappe au clavier. Ce qui est reporté ici est l'affichage des résultats (jusqu'à ce qu'ils soient prêts), et non pas les requêtes réseaux en elle-mêmes. Même si l'utilisateur continue à écrire, les réponses pour chaque frappe au clavier sont cachées (*cached, NdT*), et donc les données ne sont pas récupérées une fois de plus lorsqu'on appuie sur Espace. Cette frappe est par ailleurs instantanée.

Reviewer: ```suggestion
Remarquez qu'il y a quand même une requête réseau par frappe clavier. Ce qui est différé ici, c'est l'affichage des résultats (jusqu'à ce qu'ils soient prêts), et non pas les requêtes réseau elles-mêmes. Même si l'utilisateur continue à saisir, les réponses pour chaque frappe clavier sont mises en cache, donc les données ne sont pas chargées à nouveau lorsqu'on appuie sur Backspace : la mise à jour est alors instantanée.
```

Quelques glissements sémantiques (j'ai beaucoup aimé Backspace -> Espace 😉)

</DeepDive>

---

### Indiquer que le contenu est obsolète {/*indicating-that-the-content-is-stale*/}

> Dans l'exemple ci-dessous, il n'y aucune indication montrant que la liste des résultats est toujours en train de charger, et qu'elle prend en compte la dernière recherche. Cela peut être dérangeant pour l'utilisateur si les nouveaux résultats prennent du temps à charger. Afin que l'utilisateur comprenne que la liste de résultats n'est pas la même que celle obtenue lors de la dernière recherche, vous pouvez ajouter une indication visuelle lorsque l'ancienne liste de résultats est affichée : 

Reviewer: ```suggestion
Dans l'exemple ci-avant, il n'y aucune indication que la liste des résultats pour la dernière requête est toujours en train de charger. Cela peut être déroutant pour l'utilisateur si les nouveaux résultats prennent du temps à charger. Afin de bien signifier que la liste de résultats ne reflète pas encore la dernière recherche, vous pouvez ajouter une indication visuelle lorsque l'ancienne liste de résultats est affichée :
```

- "Above" ≠ "Below"
- Confusion sémantique
- C'est quoi ces espaces en fin de ligne à chaque fois ?

</div>
```

> Avec ce changement, dès que vous commencerez à taper, l'ancienne liste de résultats sera légèrement transparente, jusqu'à ce que la nouvelle liste de résultats soit chargée. Vous pouvez également ajouter une transition CSS pour que le rendu soit graduel, comme dans l'exemple ci-dessous : 

Reviewer: ```suggestion
Avec ce changement, dès que vous commencerez à taper, l'ancienne liste de résultats sera légèrement assombrie, jusqu'à ce que la nouvelle liste de résultats soit chargée. Vous pouvez également ajouter une transition CSS pour un résultat plus graduel, comme dans l'exemple ci-dessous :
```

---

> ### Délayer le rendu d'une partie de l'UI {/*deferring-re-rendering-for-a-part-of-the-ui*/}

Reviewer: ```suggestion
### Différer le rendu d'une partie de l'UI {/*deferring-re-rendering-for-a-part-of-the-ui*/}
```

> Vous pouvez également utiliser `useDeferredValue` pour optimiser les performances. Cette manière de faire est utile lorsqu'une partie de votre UI est lente à re-render, qu'il n'y a pas de manière facile de l'optimiser, et que vous voulez éviter qu'elle bloque le reste de l'UI.

Reviewer: ```suggestion
Vous pouvez également utiliser `useDeferredValue` pour optimiser les performances. C'est pratique lorsqu'une partie de votre UI a un rendu lent, qu'il n'y a pas de manière simple de l'optimiser, et que vous voulez éviter qu'elle bloque le reste de l'UI.
```

> Imaginez que vous avez une zone de saisie de texte et un composant (comme un graphique ou une longue liste) qui re-renders à chaque frappe sur le clavier :

Reviewer: ```suggestion
Imaginez que vous avez un champ textuel et un composant (comme un graphique ou une longue liste) qui refait son rendu à chaque frappe clavier :
```

}
```

> D'abord, optimisez `SlowList` pour éviter le re-render lorsque ses propriétés sont les mêmes. Pour se faire, [enveloppez-le dans `memo`:](/reference/react/memo#skipping-re-rendering-when-props-are-unchanged)

Reviewer: ```suggestion
Pour commencer, optimisez `SlowList` pour éviter un nouveau rendu quand ses propriétés n'ont pas changé. Pour ce faire, [enrobez-le avec `memo`](/reference/react/memo#skipping-re-rendering-when-props-are-unchanged) :
```

Trad établie pour "wrap" (on préfère "enrober" à "envelopper", tout aussi valable mais plus long, d'autant qu'on dit couramment "enrobage" aussi dans la tech).

```js {1,3}
const SlowList = memo(function SlowList({ text }) {
  // ...
});
```

> Cependant, cette manière de faire n'est utile que si les propriétés de `SlowList` sont *les mêmes* que pendant le rendu précédent. Ce composant peut toujours être lent lorsque les propriétés sont *différentes,* et lorsque vous avez besoin de montrer un visuel différent.

Reviewer: ```suggestion
Cependant, ça ne vous aide que si les propriétés de `SlowList` sont *les mêmes* que lors du rendu précédent. Ce composant peut toujours être lent lorsque les propriétés sont *différentes*, et que vous avez effectivement besoin de produire un rendu visuel distinct.
```

> Concrètement, le problème de performances est causé lorsque vous tapez dans le champ de saisie. La `SlowList` reçoit des nouvelles propriétés à chaque frappe, et l'écriture est saccadée. Son arbre entier est re-render. Dans ce cas, `useDeferredValue` vous permet de mettre à jour en priorité le champ de saisie (qui doit être rapide) plutôt que de mettre à jour la liste de résultats (qui peut être plus lente):

Reviewer: ```suggestion
Concrètement, le souci de performances principal vient de ce que lorsque vous tapez dans le champ de saisie, la `SlowList` reçoit des nouvelles propriétés, et la lenteur de sa mise à jour rend la saisie saccadée. Dans un tel cas, `useDeferredValue` vous permet de prioriser la mise à jour du champ de saisie (qui doit être rapide) par rapport à celle de la liste de résultats (qui peut être plus lente) :
```

}
```

> Cela ne rend pas le nouveau rendu de `SlowList` plus rapide. Cependant, cela dit à React que rendre la liste peut être dépriorisé afin de ne pas bloquer les frappes au clavier. La liste va "être en retard" par rapport au champ de saisie avant de le "rattraper". Comme auparavant, React va essayer de mettre à jour la liste le plus vite possible, mais sans empêcher l'utilisateur de taper.

Reviewer: ```suggestion
Ça n'accélère pas le rendu de `SlowList`. Néanmoins, ça indique à React de déprioriser le rendu de la liste afin de ne pas bloquer les frappes clavier. La liste sera « en retard » par rapport au champ de saisie, pour finalement le « rattraper ». Comme auparavant, React essaiera de mettre à jour la liste le plus vite possible, mais sans empêcher l'utilisateur de taper.
```

> <Recipes titleText="La différence entre useDeferredValue et le rendu non optimisé" titleId="examples">

Reviewer: ```suggestion
<Recipes titleText="Différence entre useDeferredValue et un rendu non optimisé" titleId="examples">
```

> #### Reporter le rendu d'une liste {/*deferred-re-rendering-of-the-list*/}

Reviewer: ```suggestion
#### Différer le rendu de la liste {/*deferred-re-rendering-of-the-list*/}
```

> Dans cet exemple, chaque élément du composant `SlowList` est **artificiellement ralenti** afin que vous puissiez observer comment `useDeferredValue` permet de garder ce champ de saisie réactif. Écrivez dans le champ de saisie et observez comment écrire semble rapide et comment la liste "est en retard".

Reviewer: ```suggestion
Dans cet exemple, chaque élément du composant `SlowList` est **artificiellement ralenti** afin que vous puissiez constater que `useDeferredValue` permet de garder le champ de saisie réactif. Écrivez dans le champ de saisie et voyez comme la saisie reste réactive, alors que la liste « est en retard ».
```

function SlowItem({ text }) {
  let startTime = performance.now();
  while (performance.now() - startTime < 1) {
    // Ne fait rien pendant 1 ms par élément pour simuler du code extrêmement lent
  }

  return (
    <li className="item">

>       Texte: {text}

Reviewer: ```suggestion
      Texte : {text}
```

<Solution />

#### Rendu non optimisé de la liste {/*unoptimized-re-rendering-of-the-list*/}

Dans cet exemple, chaque élément du composant `SlowList` est **artificiellement ralenti** sans utiliser `useDeferredValue`.

> Vous pouvez remarquer que la frappe dans le champ de saisie est très saccadée. Sans `useDeferredValue`, chaque frappe au clavier force la liste entière à re-render immédiatement et de manière inarrêtable.

Reviewer: ```suggestion
Vous pouvez remarquer que la frappe dans le champ de saisie est très saccadée. Sans `useDeferredValue`, chaque frappe clavier force la liste entière à refaire un rendu immédiatement, sans possibilité d'interruption.
```

<Pitfall>

> L'optimisation nécessite que `SlowList` soit accompagnée d'un [`memo`.](/reference/react/memo) À chaque fois que `text` change, React a besoin d'être capable de re-render le composant parent rapidement. Durant ce rendu, `deferredText` possède toujours sa valeur précédente, donc `SlowList` est capable d'ignorer le re-rendering (ses propriétés n'ont pas changé). Sans [`memo`,](/reference/react/memo) il a besoin de re-render dans tous les cas, ce qui empêche l'optimisation.

Reviewer: ```suggestion
Cette optimisation nécessite que `SlowList` soit enrobée par un [`memo`](/reference/react/memo). En effet, à chaque fois que `text` change, React doit pouvoir refaire le rendu du composant parent rapidement. Durant ce rendu, `deferredText` a toujours sa valeur précédente, donc `SlowList` peut s'épargner un nouveau rendu (ses propriétés n'ont pas changé). Sans [`memo`](/reference/react/memo), il y aurait un nouveau rendu dans tous les cas, ce qui tue tout l'intérêt de l'optimisation.
```

</Pitfall>

<DeepDive>

> #### Quelle est la différence entre reporter une valeur et le debouncing et le throttling ? {/*how-is-deferring-a-value-different-from-debouncing-and-throttling*/}

Reviewer: ```suggestion
#### Valeur différée, *debouncing* et *throttling* : quelles différences ? {/*how-is-deferring-a-value-different-from-debouncing-and-throttling*/}
```

> Il existe deux technique d'optimisation communes que vous avez pu peut-être utiliser auparavant dans ce genre de scénarios :

Reviewer: ```suggestion
Il existe deux technique d'optimisation courantes que vous avez peut-être utilisées auparavant dans ce genre de scénarios :
```

> - *Debouncing* signifie que vous attendez qu'un utilisateur arrête de taper (par exemple pendant une seconde) avant de mettre à jour la liste.

Reviewer: ```suggestion
- Le *debouncing* signifie que vous attendriez que l'utilisateur cesse de taper (par exemple pendant une seconde) avant de mettre à jour la liste.
```

> - *Throttling* signifie que vous mettez à jour la liste à un temps donné (par exemple toutes les secondes).

Reviewer: ```suggestion
- Le *throttling* signifie que vous ne mettriez à jour la liste qu'à une fréquence limitée (par exemple au maximum une fois par seconde).
```

> Mêmes si ces techniques sont utiles dans certains cas, `useDeferredValue` est plus adapté pour optimiser le rendu, car il est totalement intégré avec React et il s'adapte au support de l'utilisateur.

Reviewer: ```suggestion
Mêmes si ces techniques sont utiles dans certains cas, `useDeferredValue` est plus adapté pour optimiser le rendu, car il est totalement intégré avec React et il s'adapte à l'appareil de l'utilisateur.
```

Il ne demande pas de choisir un délai fixé, contrairement au debouncing et au throttling. Si le support de l'utilisateur est rapide (par exemple un ordinateur puissant), le rendu délayé arrive quasiment immédiatement et n'est pas visible pour l'utilisateur. Si le support de l'utilisateur est lent, la liste est "en retard" par rapport au champ de saisie, proportionnellement à la vitesse de l'appareil. 

> De plus, les rendu reportés réalisés par `useDeferredValue` sont susceptibles d'être interrompus par défaut, ce qui n'est pas le cas du debouncing ou du throttling. Cela signifie que si React est au milieu du rendu d'une liste large, mais que l'utilisateur frappe sur une autre touche, React va abandonner ce rendu, considérer la frappe, et va ensuite commencer à faire le rendu en arrière-plan encore une fois. En contraste, le debouncing et le throttling produisent toujours une expérience lente car ils *bloquent:* ils reportent simplement le moment où le rendu bloque la frappe.

Reviewer: ```suggestion
De plus, les rendus différés planifiés par `useDeferredValue` sont par défaut susceptibles d'être interrompus, ce qui n'est pas le cas du *debouncing* ou du *throttling*. Ça signifie que si React est en plein milieu du rendu d'une vaste liste, et que l'utilisateur ajuste sa saisie, React abandonnera ce rendu, traitera la frappe, et recommencera le rendu en arrière-plan. Par opposition, le *debouncing* et le *throttling* donneraient ici toujours une expérience saccadée car ils sont *bloquants* : ils diffèrent simplement le moment auquel le rendu bloque la frappe.
```

> Si vous souhaitez optimiser des éléments qui n'arrivent pas pendant le rendu, le debouncing et le throttling sont toujours utiles. Par exemple, ils peuvent vous laisser entamer moins de requêtes réseau. Vous pouvez également utiliser différentes de ces techniques réunies. 

Reviewer: ```suggestion
Si vous souhaitez optimiser des traitements hors du rendu, le *debouncing* et le *throttling* restent utiles. Par exemple, ils peuvent vous permettre de lancer moins de de requêtes réseau. Vous pouvez parfaitement combiner ces techniques. 
```
